### PR TITLE
Add asyncio events module

### PR DESCRIPTION
**Pull request by bdraco**
This is the module being used in Home Assistant and pysonos (https://github.com/amelchio/pysonos/blob/master/pysonos/events_asyncio.py) with the extras stripped out modified to work on python 3.5+.

### CODE REVIEW

**Comment by bdraco** [comment]
There is a case where I/O can happen in the event loop discovered in https://github.com/home-assistant/core/issues/48732 though which affects twisted as well.

I'll push another change to fix that but `twisted` likely needs a similar fix to run it in `twisted.internet.threads.deferToThread`

File: soco/events_asyncio.py [hunk 1]
```diff
@@ -0,0 +1,582 @@
+# pylint: disable=not-context-manager,import-error,wrong-import-position
+
+# NOTE: The pylint not-content-manager warning is disabled pending the fix of
+# a bug in pylint. See https://github.com/PyCQA/pylint/issues/782
+
```
Comment (pwt) [inline soco/events_asyncio.py:5]: This defect is closed now. Remove the disable?

File: soco/events_asyncio.py [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+# Disable while we have Python 2.x compatability
+# pylint: disable=useless-object-inheritance
+
```
Comment (pwt) [inline soco/events_asyncio.py:8]: Since SoCo 0.21 we now only support Python 3.5+. Remove comment and disable?

File: soco/events_asyncio.py [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+            await asyncio.sleep(100)
+            await before_shutdown()
+
+
+        if __name__ == "__main__":
+            asyncio.run(main())
+
+"""
+
```
Comment (pwt) [inline soco/events_asyncio.py:69]: Where does this appear in the documentation tree?

File: soco/events_asyncio.py (hunk 1 above)
Comment (bdraco) [reply soco/events_asyncio.py:5]: I used `events_twisted` as a base. I'll remove it and fix anything as a result

File: soco/events_asyncio.py (hunk 2 above)
Comment (bdraco) [reply soco/events_asyncio.py:8]: I used `events_twisted` as a base. I'll remove it and fix anything as a result

File: soco/events_asyncio.py (hunk 3 above)
Comment (bdraco) [reply soco/events_asyncio.py:62]: I made sure its included with https://github.com/SoCo/SoCo/pull/815/commits/2a1f7369992f2d535777c65c7213db9f2cd99401

File: soco/events_asyncio.py [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+if "sphinx" in sys.modules:
+
+    class Resource:  # pylint: disable=no-init
+        """Fake Resource class to use when building docs"""
+
+
+else:
+    import asyncio
+    from aiohttp import ClientSession, web
+
```
Comment (pwt) [inline soco/events_asyncio.py:78]: This hack doesn't appear to be necessary any more?

**Comment by pwt** [comment]
Hi @bdraco:

Thanks for this great contribution. I'm optimistically targeting its inclusion in the upcoming v0.22 release (17th April).

I'm going to continue to review and make comments. Please excuse me making comments in multiple passes ... I only have a superficial knowledge of asyncio.

File: soco/events_asyncio.py [hunk 5]
```diff
@@ -0,0 +1,575 @@
...
+            await asyncio.sleep(100)
+            await before_shutdown()
+
+
+        if __name__ == "__main__":
+            asyncio.run(main())
+
+"""
+
+import sys
```
Comment (pwt) [inline soco/events_asyncio.py:63]: Import not required.

File: soco/events_asyncio.py [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+
+The `Subscription` class from this module will be used in
+:py:mod:`soco.services` if `config.EVENTS_MODULE` is set
+to point to this module.
+
+Example:
+
+    Run this code, and change your volume, tracks etc::
+
+        from __future__ import print_function
```
Comment (pwt) [inline soco/events_asyncio.py:13]: Not required for Python 3.

File: soco/events_asyncio.py [hunk 7]
```diff
@@ -0,0 +1,575 @@
...
+# Hack to make docs build without asyncio installed
+if "sphinx" in sys.modules:
+
+    class Resource:  # pylint: disable=no-init
+        """Fake Resource class to use when building docs"""
+
+
+else:
+    import asyncio
+    from aiohttp import ClientSession, web
```
Comment (pwt) [inline soco/events_asyncio.py:77]: New requirement `aiohttp` needs to be added to requirements.txt.

File: soco/events_asyncio.py [hunk 8]
```diff
@@ -0,0 +1,575 @@
...
+            asyncio.run(main())
+
+"""
+
+import sys
+import logging
+import socket
+import time
+
+# Hack to make docs build without asyncio installed
```
Comment (pwt) [inline soco/events_asyncio.py:68]: This hack doesn't appear to be necessary any more?

File: soco/events_asyncio.py [hunk 9]
```diff
@@ -0,0 +1,575 @@
+# pylint: disable=import-error,wrong-import-position
```
Comment (pwt) [inline soco/events_asyncio.py:1]: Doesn't seem to be required.

**Comment by pwt** [comment]
I've pushed the changes I proposed in my review to your branch (https://github.com/SoCo/SoCo/pull/815/commits/a0303a6c1a0ce9482de3721aced9c332f376909a). Feel free to review and revert, of course.

**Comment by bdraco** [comment]
> I've pushed the changes I proposed in my review to your branch (https://github.com/SoCo/SoCo/pull/815/commits/a0303a6c1a0ce9482de3721aced9c332f376909a). Feel free to review and revert, of course.

Nice! LGTM 👍

**Comment by pwt** [comment]
I think this is ready to merge, but I wouldn't mind another set of eyes on the changes to `events_base.py` in particular.  (@KennethNielsen if you have time perhaps you could take a quick look. No problem if not.)

In the absence of further comments, I plan to merge this in the next few days, for inclusion in v0.22.

**Comment by bdraco** [comment]
Thanks!

**Pull request merged**